### PR TITLE
Release mode toggle

### PR DESCRIPTION
**Pull request by akashkj**
## Product Description
Change in the UI of release manager to disable release buttons by default. Behavior is applicable for new domains and optional for existing domains.

image

Project settings
image

## Technical Summary
https://docs.google.com/document/d/1D6eKnNIonr0XzZu-hLAwG-ftWmotwc0rZGCGpoS2Pkc/edit

## Feature Flag
NA

### Safety story
Tested on staging

### Automated test coverage

Added unit test for project settings interface.

### QA Plan
https://dimagi-dev.atlassian.net/browse/QA-4550

### Migrations
- [x] The migrations in this code can be safely applied first independently of the code

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/apps/domain/forms.py
Comment (mkangia) [inline corehq/apps/domain/forms.py:503]: We had to add a new model for this! 
Can't we store this on domain itself?

Comment (akashkj) [reply corehq/apps/domain/forms.py:503]: There was a review comment in one of the previous PR to avoid adding new attributes to `Domain` class and rather create a separate model or each new functionality.

**Review by orangejenny — CHANGES_REQUESTED** [review]
Left a couple of comments on coding best practices. At a high level, I have concerns about the design here and left some comments on the spec [here](https://docs.google.com/document/d/1D6eKnNIonr0XzZu-hLAwG-ftWmotwc0rZGCGpoS2Pkc/edit?disco=AAAAd78gji8). Marking this PR request changes for now, because at least one of the spec comments is blocking (a domain-level setting shouldn't be controlled on an app-specific page).

File: corehq/apps/domain/migrations/0014_appreleasemodesetting.py
Comment (mkangia) [reply corehq/apps/domain/migrations/0014_appreleasemodesetting.py:1]: Looks like this is already deployed on staging, so in that case its better to just leave it as it is?

Comment (orangejenny) [reply corehq/apps/domain/migrations/0014_appreleasemodesetting.py:1]: I'd still rather squash them. It's not a lot of effort to roll these back on staging, squash, and then redeploy. If they're unsquashed, it's a little bit of extra overhead (and code complexity - the `auto` migration isn't named well) that we'll never get rid of.

File: corehq/apps/domain/forms.py
Comment (orangejenny) [reply corehq/apps/domain/forms.py:503]: yeah, this should be a new model (it's likely I'm the person who left the comment on the previous PR). It's not directly because of couch to sql (although I do think there's potential to split up the domain model as part of that migration), it's more just avoiding increasing the size of the domain model, which already does way too many things.

File: corehq/apps/app_manager/static/app_manager/js/releases/app_view_release_manager.js
Comment (mkangia) [reply corehq/apps/app_manager/static/app_manager/js/releases/app_view_release_manager.js:23]: Hey @orangejenny 
will let @akashkj any details I missed.

From the design perspective, the toggle looked better to be above the releases table and hence the way javascript happened. We did try it next to "Show only released versions" first but that threw other elements off and having toggles icons/buttons next to a text toggle to show released versions wasn't looking pleasing as well.
It made sense to keep it on top where its visible.

Feel free to play around with HTML, quick link to an app [here](https://staging.commcarehq.org/a/fhir/apps/view/d03d683c180149c4b5f54b8ba6ebcfca/). Things did become harder when we got that toggle outside the table, so if we can move it back, it would be less code only.

File: corehq/apps/app_manager/views/view_generic.py
Comment (mkangia) [inline corehq/apps/app_manager/views/view_generic.py:346]: nit: this variable `release_control_visible` seems redundant here.

File: corehq/apps/domain/migrations/0014_appreleasemodesetting_squashed_0015_auto_20220930_2012.py
Comment (mkangia) [inline corehq/apps/domain/migrations/0014_appreleasemodesetting_squashed_0015_auto_20220930_2012.py:8]: was `0014_appreleasemodesetting` added by this PR?

**Comment by mkangia** [comment]
https://github.com/dimagi/commcare-hq/pull/32245#pullrequestreview-1171422066
@orangejenny were your concerns with the design concluded on the doc? are we good to proceed with this implementation

**Comment by orangejenny** [comment]
@mkangia didn't have a chance to look at this today but should be able to do so tomorrow

File: corehq/apps/domain/migrations/0014_appreleasemodesetting_squashed_0015_auto_20220930_2012.py
Comment (akashkj) [reply corehq/apps/domain/migrations/0014_appreleasemodesetting_squashed_0015_auto_20220930_2012.py:8]: Yes

File: corehq/apps/domain/models.py
Comment (akashkj) [reply corehq/apps/domain/models.py]: Removed

File: corehq/apps/app_manager/templates/app_manager/app_view_release_manager.html
Comment (akashkj) [reply corehq/apps/app_manager/templates/app_manager/app_view_release_manager.html]: Added.

File: corehq/apps/domain/migrations/0014_appreleasemodesetting.py
Comment (akashkj) [reply corehq/apps/domain/migrations/0014_appreleasemodesetting.py]: Merged the migrations.

File: corehq/apps/app_manager/views/view_generic.py
Comment (akashkj) [reply corehq/apps/app_manager/views/view_generic.py:346]: It is used here - https://github.com/dimagi/commcare-hq/pull/32245/files#diff-cf519579ec6065016857e65ebdc75d47106708c1bb2022dbb47f14d927395b8aR44

File: corehq/apps/domain/migrations/0014_appreleasemodesetting_squashed_0015_auto_20220930_2012.py
Comment (mkangia) [reply corehq/apps/domain/migrations/0014_appreleasemodesetting_squashed_0015_auto_20220930_2012.py:8]: But its not showing in the PR?
If its meant to replace the migrations, they would still need to be present so that django does not complain about missing migrations.
If you were just merging them to make a single migration for the PR, you don't need to add replaces here.

**Comment by mkangia** [comment]
> didn't have a chance to look at this today but should be able to do so tomorrow

that works.

Also, this one,
https://github.com/dimagi/commcare-hq/pull/32245#discussion_r1037984609

These two are design questions that would be good to unblock on.

**Review by mkangia — CHANGES_REQUESTED** [review]

File: corehq/apps/app_manager/static/app_manager/js/releases/releases.js
Comment (orangejenny) [inline corehq/apps/app_manager/static/app_manager/js/releases/releases.js:509]: Nitpick: is this needed? Is the intention that other code not access observables directly?

File: corehq/apps/domain/forms.py
Comment (orangejenny) [inline corehq/apps/domain/forms.py:427]: A link to docs or a longer explanation would be helpful.

**Review by orangejenny — APPROVED** [review]

**Comment by orangejenny** [comment]
@akashkj Could you add a screenshot (or more than one) of the final UI to the PR description?

File: corehq/apps/domain/migrations/0014_appreleasemodesetting_squashed_0015_auto_20220930_2012.py
Comment (akashkj) [reply corehq/apps/domain/migrations/0014_appreleasemodesetting_squashed_0015_auto_20220930_2012.py:8]: Replaced with a new migration.

**Comment by akashkj** [comment]
> @akashkj Could you add a screenshot (or more than one) of the final UI to the PR description?

Added

File: corehq/apps/app_manager/static/app_manager/js/releases/releases.js
Comment (akashkj) [reply corehq/apps/app_manager/static/app_manager/js/releases/releases.js:509]: Yes, that's the intention.

File: corehq/apps/domain/forms.py
Comment (mkangia) [reply corehq/apps/domain/forms.py]: I think we should add a confluence page for this since its a public facing feature and link that here.
@orangejenny I am guessing this would be the recommended usual path anyway for user facing features?

File: corehq/apps/domain/migrations/0014_appreleasemodesetting_squashed_0015_auto_20220930_2012.py
Comment (mkangia) [reply corehq/apps/domain/migrations/0014_appreleasemodesetting_squashed_0015_auto_20220930_2012.py]: thank you for fixing this @akashkj

Comment (mkangia) [reply corehq/apps/domain/migrations/0014_appreleasemodesetting_squashed_0015_auto_20220930_2012.py]: what was needed for this to be done on staging?

File: corehq/apps/domain/forms.py
Comment (orangejenny) [reply corehq/apps/domain/forms.py]: yeah, that's a good point - the new description is helpful, **and** linking to docs is also always a good idea

**Comment by orangejenny** [comment]
@akashkj thanks!

File: corehq/apps/domain/migrations/0014_appreleasemodesetting_squashed_0015_auto_20220930_2012.py
Comment (akashkj) [reply corehq/apps/domain/migrations/0014_appreleasemodesetting_squashed_0015_auto_20220930_2012.py]: No change done on staging.

File: corehq/apps/domain/forms.py
Comment (mkangia) [inline corehq/apps/domain/forms.py:632]: what is this doing?

File: corehq/apps/app_manager/static/app_manager/js/releases/releases.js
Comment (mkangia) [inline corehq/apps/app_manager/static/app_manager/js/releases/releases.js:202]: why do we need to lock things when someone clicks on "Publish"?

Comment (mkangia) [reply corehq/apps/app_manager/static/app_manager/js/releases/releases.js:202]: Actually this is erroring since `releaseLocked` is not present on this object.

File: corehq/apps/app_manager/static/app_manager/js/releases/releases.js
Comment (mkangia) [inline corehq/apps/app_manager/static/app_manager/js/releases/releases.js:381]: is this being used anywhere?

File: corehq/apps/app_manager/static/app_manager/js/releases/releases.js
Comment (mkangia) [inline corehq/apps/app_manager/static/app_manager/js/releases/releases.js:382]: what is the difference between releasesEnabled and releaseLocked. By looking at their usages I really can't tell their purpose.

File: corehq/apps/app_manager/static/app_manager/js/releases/releases.js
Comment (akashkj) [reply corehq/apps/app_manager/static/app_manager/js/releases/releases.js:381]: No, it is not used now. Removed.

File: corehq/apps/domain/forms.py
Comment (akashkj) [reply corehq/apps/domain/forms.py:632]: Not used now. It was used earlier to check whether to hide this setting, once enabled.

File: corehq/apps/app_manager/static/app_manager/js/releases/releases.js
Comment (akashkj) [reply corehq/apps/app_manager/static/app_manager/js/releases/releases.js]: We are using `releasesEnabled` only now, removed the references for `releaseLocked`

File: corehq/apps/app_manager/static/app_manager/js/releases/releases.js
Comment (akashkj) [reply corehq/apps/app_manager/static/app_manager/js/releases/releases.js]: Cleaned up `releaseLocked`

File: corehq/apps/app_manager/static/app_manager/js/releases/app_view_release_manager.js
Comment (mkangia) [inline corehq/apps/app_manager/static/app_manager/js/releases/app_view_release_manager.js:83]: nit: Feel free to ignore this if don't feel worth it. I think I shared this feedback earlier as well.
This would be better named as "showReleaseControl".

File: corehq/apps/domain/forms.py
Comment (mkangia) [inline corehq/apps/domain/forms.py:433]: We should also add that,

Once enabled, it cannot be changed/disabled. 
(whatever the wording seems better)

File: corehq/apps/domain/migrations/0014_appreleasemodesetting.py
Comment (mkangia) [inline corehq/apps/domain/migrations/0014_appreleasemodesetting.py:17]: :+1: index

**Review by mkangia — COMMENTED** [review]
The conflicts with master seem to be messing up the commit history quite a bit.
Can you reword the last commit? I think its just a "q" before the merge message.

Plus, I noted in review that we should inform the user that they can't undo the setting change once enabled. It would throw them by surprise if they enable to just see what the change is and then not be able to see the setting entirely, in case they decide to revert to old implementation.

**Comment by mkangia** [comment]
So, this is now waiting on documentation?

File: corehq/apps/domain/forms.py
Comment (akashkj) [reply corehq/apps/domain/forms.py:433]: "Once this setting is enabled, it cannot be disabled" - sounds good?

File: corehq/apps/app_manager/static/app_manager/js/releases/app_view_release_manager.js
Comment (akashkj) [reply corehq/apps/app_manager/static/app_manager/js/releases/app_view_release_manager.js:83]: Ok. I think you are referring to this comment from earlier? https://github.com/dimagi/commcare-hq/pull/32245#discussion_r1013287887

**Comment by akashkj** [comment]
> So, this is now waiting on documentation?

There is a pending fix for a bug that Ashish had reported where button are visible without any versions. Still to verify the complete code changes for it.

File: corehq/apps/app_manager/static/app_manager/js/releases/app_view_release_manager.js
Comment (mkangia) [reply corehq/apps/app_manager/static/app_manager/js/releases/app_view_release_manager.js]: Ya, `showReleaseControls` is even better than `setReleaseLockButtons`

File: corehq/apps/domain/forms.py
Comment (orangejenny) [reply corehq/apps/domain/forms.py:433]: works for me

**Comment by akashkj** [comment]
@mkangia Added changes based on change in requirements.

File: corehq/apps/domain/forms.py
Comment (mkangia) [inline corehq/apps/domain/forms.py:627]: Can you share the reason for this change?

**Comment by mkangia** [comment]
@akashkj changes look good.
Is this going through QA?

Can do a final review pass once QA is done and the toggles being visible for no releases is fixed.

File: corehq/apps/domain/forms.py
Comment (akashkj) [reply corehq/apps/domain/forms.py:627]: To save object again only in case of change from existing.

**Comment by akashkj** [comment]
@mkangia Yes, QA needed again.
Two pending changes:
- Toggle visibility
- Icons' size consistency

**Comment by akashkj** [comment]
Added fix for alignment issue(https://github.com/dimagi/commcare-hq/pull/32245/commits/e984de994f1a9cab4e2dbc18c4e5d29e451cb9a2):
image
With fix:
image

image

**Comment by akashkj** [comment]
Recent change introduced in the functionality:
- Release mode always enabled for new domains and optional for existing domains is changed to optional for all the domains with default being disabled
- Once enabled in project settings, it was not allowed to disable the release mode. This is changed to always allow user to change this setting with no restriction.

Bug fixes:
- Lock and unlock icons having different width (https://github.com/dimagi/commcare-hq/pull/32245/commits/93204743e360ae1892376a2c3e65268158a8698b)
- Release mode buttons were visible even if no version was created for app (https://github.com/dimagi/commcare-hq/pull/32245/commits/abd873b5eabfd006595b715815c1fe1f9640c4cf)
- In community plan, due to difference in UI structure where tabbed elements were not present, alignment was broken (https://github.com/dimagi/commcare-hq/pull/32245/commits/e984de994f1a9cab4e2dbc18c4e5d29e451cb9a2)

@orangejenny Please check if changes look fine.

fyi @mkangia

**Comment by orangejenny** [comment]
thank you for the alignment fix - appreciate the attention to detail

File: corehq/apps/app_manager/static/app_manager/js/releases/releases.js
Comment (mkangia) [reply corehq/apps/app_manager/static/app_manager/js/releases/releases.js:252]: apologies for stressing on naming again.
can we think of better name for `releasesEnabled`, like `allowReleaseOperations`/`showReleaseOperations`?

Comment (mkangia) [reply corehq/apps/app_manager/static/app_manager/js/releases/releases.js:252]: I always thought `showReleaseControl` was to show or hide the lock buttons but its enabling/disabling the release buttons.

so, `showReleaseControl` and `releasesEnabled` are doing the same thing right? just that one is a function and other is an attribute that we are setting.
I think then this could be the same name for both?

File: corehq/apps/app_manager/static/app_manager/js/releases/app_view_release_manager.js
Comment (mkangia) [inline corehq/apps/app_manager/static/app_manager/js/releases/app_view_release_manager.js:83]: From the other comment, if we are keeping `releasesEnabled` false by default now, this would not be needed.

File: corehq/apps/app_manager/static/app_manager/js/releases/app_view_release_manager.js
Comment (mkangia) [inline corehq/apps/app_manager/static/app_manager/js/releases/app_view_release_manager.js:96]: when we click on unlock, we are passing `false` to `showReleaseControl`. Should not this be opposite?

**Review by mkangia — COMMENTED** [review]
Looks good, just a couple of queries/confirmations & a naming suggestion to make things easier from code perspective.

File: corehq/apps/app_manager/static/app_manager/js/releases/app_view_release_manager.js
Comment (akashkj) [reply corehq/apps/app_manager/static/app_manager/js/releases/app_view_release_manager.js:96]: `btn-release-unlocked` is linked to unlock icon and is used to display the current status of the screen i.e. when this button is clicked, releases are already enabled and to be disabled.

Comment (mkangia) [reply corehq/apps/app_manager/static/app_manager/js/releases/app_view_release_manager.js]: Right!
Thank you!

File: corehq/apps/app_manager/static/app_manager/js/releases/releases.js
Comment (mkangia) [inline corehq/apps/app_manager/static/app_manager/js/releases/releases.js]: :+1:

File: corehq/apps/domain/tests/test_forms.py
Comment (mkangia) [inline corehq/apps/domain/tests/test_forms.py:211]: where there more tests planned or there were more tests earlier using this function?
I see only one use now, which passed the domain object and does not have any more kwargs, so looks like this could be simplified? You probably could remove this function now entirely

**Review by mkangia — APPROVED** [review]
Much thanks for the work on this @akashkj.
Appreciate adjusting to the change in requirements & addressing multiple reviews.

File: corehq/apps/domain/tests/test_forms.py
Comment (akashkj) [reply corehq/apps/domain/tests/test_forms.py]: Yes, tests that were not valid anymore were removed.
https://github.com/dimagi/commcare-hq/pull/32245/commits/ee34ec98d1d5784da48e31f98de77fc6d62547c7

**Comment by akashkj** [comment]
Merged based on last successful build with https://github.com/dimagi/commcare-hq/pull/32245/commits/78436baa9df1bd4e31c4d5ea20d7b3ad8fa6c4fb and no new change added. Stickler failing repeatedly with cloning timeout error.

**Pull request merged**
